mex.go

```diff
@@ -43,28 +46,85 @@ const (
 	mexChannelBufferSize = 2
 )
 
+type errNotifier struct {
+	c   chan struct{}
+	err error
+}
+
+func newErrNotifier() errNotifier {
+	return errNotifier{c: make(chan struct{})}
+}
+
+// Notify will store the error and notify all waiters on c that there's an error.
+func (e *errNotifier) Notify(err error) error {
+	// The code should never try to Notify(nil).
+	if err == nil {
+		panic("cannot Notify with no error")
+	}
+
+	// There may be some sort of race where we try to notify the mex twice.
+	select {
+	case <-e.c:
+		return fmt.Errorf("cannot broadcast error: %v, already have: %v", err, e.err)
+	default:
+	}
```

Comment:
Does this guard actually protect us from races? It seems possible for concurrent callers to pass through the select statement together and reach line 72 at the same time, potentially over-writing each other.

Perhaps a double-checked lock would be more robust?